Add runtime DataTemplate updates and auto-refresh for presenters (DEV TOOLING)

carldebilly opened the pull request:
This PR introduces a runtime DataTemplate update system for the Uno Platform that enables dynamic modification of DataTemplate factories and automatic refresh of controls using those templates. This is designed as a development/debugging tool to support scenarios like hot reload.

- Adds `TemplateManager` with methods to enable subscriptions and update DataTemplate factories
- Implements a subscription system allowing controls to listen for template updates and refresh automatically
- Updates built-in controls (ContentPresenter, ContentControl, ItemsControl, ItemsRepeater) to support dynamic template updates
- Includes linker optimizations to exclude the feature from production builds by default (in applications using Uno)

# Feature

## What is the new behavior? 🚀
This PR introduces runtime DataTemplate factory updates with auto-refresh capability for controls, designed as a debugging tool. The system allows modifying DataTemplate factories at runtime and automatically notifies subscribed controls to refresh their content.

## PR Checklist ✅

Please check if your PR fulfills the following requirements:

- [X] 📝 Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.
- [ ] 🧪 Added [Runtime tests, UI tests, or a manual test sample](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [X] 📚 Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] 🖼️ Validated PR `Screenshots Compare Test Run` results.
- [X] ❗ Contains **NO** breaking changes

carldebilly committed:
feat: Add runtime DataTemplate updates and auto-refresh for presenters

Introduced support for DataTemplate factory updates at runtime with a subscription mechanism. Updated related controls (ContentPresenter, ItemsControl, ContentControl) to handle template changes and added a sample demonstrating this functionality.

unodevops commented:
🤖 Your Docs stage site is ready! Visit it here: https://unodocsprstaging.z13.web.core.windows.net/pr-21342/docs/index.html

unodevops commented:
🤖 Your WebAssembly Skia Sample App stage site is ready! Visit it here: https://unowasmprstaging.z20.web.core.windows.net/pr-21342/wasm-skia-net9/index.html

carldebilly committed:
feat: Add support for dynamic template updates in ItemsRepeater

Implemented runtime subscription mechanism for DataTemplate updates in `ItemsRepeater`. Enhanced template reset logic to support conditional behavior based on update feature status. Added sample to demonstrate dynamic updates.

carldebilly committed:
feat: Add ListView sample in DataTemplate update scenario

carldebilly committed:
feat: Add conditional compilation for DataTemplate updates in sample app

carldebilly committed:
refactor: Rename `DataTemplateHelper` to `TemplateManager` to avoid conflicts

carldebilly committed:
feat: Add TemplateManager documentation and enhanced API for runtime DataTemplate updates

carldebilly committed:
refactor: Consolidate dynamic template update logic in `ItemsRepeater`

Streamlined dynamic DataTemplate update handling by removing redundant code and consolidating logic into dedicated methods.

carldebilly committed:
refactor: Make the new code in ItemsRepeater less intrusive into existing code.

unodevops commented:
🤖 Your Docs stage site is ready! Visit it here: https://unodocsprstaging.z13.web.core.windows.net/pr-21342/docs/index.html

carldebilly committed:
docs: Update TemplateManager documentation for improved clarity and advanced usage

dr1rrb commented on src/Uno.UI/UI/Xaml/Controls/Repeater/ItemsRepeater.cs:
```diff
@@ -752,30 +754,40 @@ void OnItemTemplateChanged(object oldValue, object newValue)
-						if (GetVirtualizationInfo(child).IsRealized)
+						// Walk through all the elements and make sure they are cleared for
+						// non-virtualizing layouts.
+						foreach (var element in Children)
```
This is from WinUI source code, we should not rename it

dr1rrb commented on src/Uno.UI/UI/Xaml/Controls/Repeater/ItemsRepeater.cs:
```diff
@@ -826,6 +838,125 @@ void OnItemTemplateChanged(object oldValue, object newValue)
 			InvalidateMeasure();
 		}
 
+		#region Dynamic Template Update Support
```
I would move this to a dedicated file

dr1rrb commented on src/Uno.UI/UI/Xaml/Controls/Repeater/ItemsRepeater.cs:
```diff
@@ -826,6 +838,125 @@ void OnItemTemplateChanged(object oldValue, object newValue)
+		/// </summary>
+		private void RefreshAllItems()
+		{
+			if (Layout is not { } layout)
+			{
+				return;
+			}
+
+			var args = new NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction.Reset);
+			using var processingChange = Disposable.Create(() => m_processingItemsSourceChange = null);
+			m_processingItemsSourceChange = args;
+
+			if (layout is VirtualizingLayout virtualLayout)
+			{
+				// Pass ItemTemplate (the current template) because in a dynamic update scenario,
+				// the template instance hasn't changed - only its content has been modified.
+				virtualLayout.OnItemsChangedCore(GetLayoutContext(), ItemTemplate, args);
+			}
+			else if (layout is NonVirtualizingLayout nonVirtualLayout)
+			{
+				// Walk through all the elements and make sure they are cleared for
+				// non-virtualizing layouts.
+				foreach (var child in Children)
+				{
+					if (GetVirtualizationInfo(child).IsRealized)
+					{
+						ClearElementImpl(child);
+					}
+				}
+			}
```
Unless I missed something, this is a duplicate of `TriggerTemplateReset`

dr1rrb commented on src/Uno.UI/UI/Xaml/Controls/Repeater/RecyclePool.cs:
```diff
@@ -36,6 +36,11 @@ public UIElement TryGetElement(string key)
 		public UIElement TryGetElement(string key, UIElement owner)
 			=> TryGetElementCore(key, owner);
 
+		internal void Clear()
+		{
+			m_elements.Clear();
+		}
```
Please flag uno specific code in files imported from WinUI, eg:

```suggestion
#if HAS_UNO
		// Uno specific: Helper used by template updater
		internal void Clear()
		{
			m_elements.Clear();
		}
#endif
```

dr1rrb commented on src/Uno.UI/UI/Xaml/FrameworkTemplate.cs:
```diff
@@ -196,6 +195,34 @@ public bool Equals(FrameworkTemplate? left, FrameworkTemplate? right) =>
+		// A special mode is required for it to work and is activated directly in the Uno.TemplateManager.
+
+		var previous = _viewFactory;
+		var newFactory = factory?.Invoke(previous);			if (newFactory != previous)
```
```suggestion
		var newFactory = factory?.Invoke(previous);
		if (newFactory != previous)
```

dr1rrb commented on src/Uno.UI/Uno/TemplateManager.cs:
```diff
@@ -0,0 +1,88 @@
+using View = Microsoft.UI.Xaml.UIElement;
+#endif
+
+namespace Uno
```
I would have put this in the `Uno.UI` namespace

carldebilly committed:
chore: Fixed IDE0055 compilation issues

unodevops commented:
🤖 Your WebAssembly Skia Sample App stage site is ready! Visit it here: https://unowasmprstaging.z20.web.core.windows.net/pr-21342/wasm-skia-net9/index.html

unodevops commented:
🤖 Your Docs stage site is ready! Visit it here: https://unodocsprstaging.z13.web.core.windows.net/pr-21342/docs/index.html

carldebilly committed:
refactor: Update namespaces for TemplateManager and related classes

`Uno.TemplateManager` -> `Uno.UI.TemplateManager`

unodevops commented:
⚠️⚠️ The build [174533](https://dev.azure.com/uno-platform//Uno%20Platform/_build/results?buildId=174533) has failed on Uno.UI - CI.

carldebilly committed:
refactor: Extract Uno-specific template update logic into partial class

Simplifies maintenance by isolating Uno-specific logic.

unodevops commented:
⚠️⚠️ The build [174555](https://dev.azure.com/uno-platform//Uno%20Platform/_build/results?buildId=174555) has failed on Uno.UI - CI.

carldebilly committed:
refactor: Replace IDisposable-based template subscriptions with owner-based API

Streamlines dynamic DataTemplate update handling by eliminating member disposables in favor of owner-associated subscriptions.

carldebilly committed:
feat: Add support for conditional dynamic DataTemplate updates

Introduced `TemplateManager.IsDataTemplateDynamicUpdateEnabled` to conditionally enable dynamic template updates, optimized for debugging scenarios. Enhanced configuration through MSBuild and runtime options. Updated related documentation and added guards to prevent unintended usage in production builds.

carldebilly commented on src/Uno.UI/UI/Xaml/Controls/ContentControl/ContentControl.cs:
```diff
@@ -337,11 +337,25 @@ public void UpdateContentTemplateRoot()
+
 				//Only apply template if it has changed
-				if (!object.Equals(dataTemplate, _dataTemplateUsedLastUpdate))
+				if (!object.Equals(dataTemplate, _dataTemplateUsedLastUpdate) || templateCanBeUpdated)
```
This case will only happend in a very special situation where performance is not required.

unodevops commented:
🤖 Your Docs stage site is ready! Visit it here: https://unodocsprstaging.z13.web.core.windows.net/pr-21342/docs/index.html

carldebilly committed:
refactor: Consolidate dynamic DataTemplate update logic across multiple controls

unodevops commented:
🤖 Your Docs stage site is ready! Visit it here: https://unodocsprstaging.z13.web.core.windows.net/pr-21342/docs/index.html

unodevops commented:
🤖 Your WebAssembly Skia Sample App stage site is ready! Visit it here: https://unowasmprstaging.z20.web.core.windows.net/pr-21342/wasm-skia-net9/index.html

carldebilly committed:
ci: Fixed build

unodevops commented:
🤖 Your Docs stage site is ready! Visit it here: https://unodocsprstaging.z13.web.core.windows.net/pr-21342/docs/index.html

carldebilly committed:
chore: Apply copilot suggestions

carldebilly committed:
docs: Update TemplateManager documentation

unodevops commented:
🤖 Your Docs stage site is ready! Visit it here: https://unodocsprstaging.z13.web.core.windows.net/pr-21342/docs/index.html

carldebilly commented on build/nuget/uno.winui.common.targets:
```diff
@@ -11,12 +11,17 @@
 	<!-- Feature configuration -->
 	<PropertyGroup>
 		<UnoDragDropExternalSupport Condition="'$(UnoDragDropExternalSupport)'==''">true</UnoDragDropExternalSupport>
+		<UnoEnableDynamicDataTemplateUpdate Condition="'$(UnoEnableDynamicDataTemplateUpdate)'==''">false</UnoEnableDynamicDataTemplateUpdate>
```
```suggestion
		<UnoEnableDynamicDataTemplateUpdate Condition="'$(UnoEnableDynamicDataTemplateUpdate)'==''">true</UnoEnableDynamicDataTemplateUpdate>
```
BUT: If $(Optimize) is "true", it should be false instead...

carldebilly committed:
feat: Add unavailable feature notice for WinAppSDK in DataTemplate_Update_Sample

carldebilly committed:
feat: Enable dynamic DataTemplate updates by default in user applications.

carldebilly committed:
chore: Finish WinAppSDK unavailability in DataTemplate_Update_Sample

unodevops commented:
🤖 Your Docs stage site is ready! Visit it here: https://unodocsprstaging.z13.web.core.windows.net/pr-21342/docs/index.html

jeromelaban commented on src/Uno.UI/UI/Xaml/Controls/ItemsControl/ItemsControl.cs:
```diff
@@ -186,8 +186,19 @@ public DataTemplate ItemTemplate
 		{
-			Refresh();
-			UpdateItems(null);
+			void OnCurrentItemTemplateUpdated()
+			{
+				// Recreate item containers and refresh items
+				Refresh();
+				UpdateItems(null);
+			}
+
+			if (TemplateManager.IsDataTemplateDynamicUpdateEnabled)
+			{
```
```suggestion
			if (TemplateManager.IsDataTemplateDynamicUpdateEnabled)
			{
				  void OnCurrentItemTemplateUpdated()
			      {
				      // Recreate item containers and refresh items
				      Refresh();
				      UpdateItems(null);
			      }


```

carldebilly commented on src/Uno.UI/Uno/TemplateManager.cs:
```diff
@@ -0,0 +1,147 @@
+	/// Manager for runtime DataTemplate updates and notification system.
+	/// Also exposes a global flag to enable template-update subscriptions in controls.
+	/// </summary>
+	public static class TemplateManager
```
@jeromelaban, are you comfortable with the idea of adding HD as "internals visible to" from Uno?

unodevops commented:
⚠️⚠️ The build [174704](https://dev.azure.com/uno-platform//Uno%20Platform/_build/results?buildId=174704) has failed on Uno.UI - CI.

carldebilly committed:
chore: Adjust code per PR review

carldebilly commented on src/Uno.UI/UI/Xaml/FrameworkTemplate.cs:
```diff
@@ -30,11 +30,19 @@ namespace Microsoft.UI.Xaml
+		/// scenarios (e.g., hot reload, design-time updates). Outside of this mode, the field
+		/// should remain unchanged. See Uno.UI.TemplateManager for details.
+		/// </summary>
+		internal NewFrameworkTemplateBuilder? _viewFactory;
```
There's no setting outside of the class, it's just read by FrameworkTemplatePool and it's probably using a field directly for performance reasons ?

unodevops commented:
🤖 Your Docs stage site is ready! Visit it here: https://unodocsprstaging.z13.web.core.windows.net/pr-21342/docs/index.html

carldebilly committed:
chore: Clean up comments and whitespace in dynamic DataTemplate update logic

unodevops commented:
🤖 Your WebAssembly Skia Sample App stage site is ready! Visit it here: https://unowasmprstaging.z20.web.core.windows.net/pr-21342/wasm-skia-net9/index.html

unodevops commented:
🤖 Your Docs stage site is ready! Visit it here: https://unodocsprstaging.z13.web.core.windows.net/pr-21342/docs/index.html

unodevops commented:
🤖 Your WebAssembly Skia Sample App stage site is ready! Visit it here: https://unowasmprstaging.z20.web.core.windows.net/pr-21342/wasm-skia-net9/index.html
